Reader: Fix crash in site stream and bug leading to data loss in tag stream

dvdchr opened the pull request:
Fixes #22165

This PR fixes two issues in the Reader:

1. As described in https://github.com/wordpress-mobile/WordPress-iOS/issues/22165#issuecomment-2134116245, when navigating to the same site stream multiple times, the site topic could be deleted, causing the post objects to be faulted when they're still being displayed in the previous screens. Any access to the faulted object's properties raises an exception.
    - **Root cause:** When requesting for a `ReaderSiteTopic`, any existing site topic having the same site ID gets deleted from context. The problem is, `ReaderSiteTopic` has a cascade delete relationship with posts, which turns all post objects in the previous screens into a fault.
    - https://github.com/wordpress-mobile/WordPress-iOS/commit/a94ca1fafe516d3e6fbbde32dba0fa2a3aee09f2 solves this by not deleting the existing site topic whenever we're requesting a new site topic. If there's an existing one, the method `[self siteTopicForRemoteSiteInfo:siteInfo inContext:context];` already performs an upsert. This means that if we have several site streams (for the same site) on the navigation stack, they will all reference the same site topic.

2. When performing a pull-to-refresh on the tag stream, the service wipes out all posts associated with the tag, and this could also delete saved posts (i.e., data loss).
    - **Root cause:** `ReaderPostStreamService` is used to fetch data for tag streams. When fetching data for the first page, the service wipes all data related to the tag, but there's no check in place for posts that are in use or saved.
    - https://github.com/wordpress-mobile/WordPress-iOS/commit/62703dae563cd560babaf239b8291448d9ed0896 solves this by skipping posts that are in use (`inUse == true`) or saved (`isSavedForLater == true`) when deleting posts associated to the tag. The fetch operation merges the new remote object with existing objects based on the `globalID` property (see `ReaderPost.createOrReplace`). Since `inUse` and `isSavedForLater` is purely a client-side property, it won't be overridden by the values from remote.

## To test

Run the unit tests and ensure they pass 🟢 . In addition, here are the manual testing steps for each scenario:

### Case 1: Crash when previewing site

Follow the reproduction steps described in https://github.com/wordpress-mobile/WordPress-iOS/issues/22165#issuecomment-2134116245:

1. Launch the Jetpack app
2. Navigate to Reader
3. From the Discover stream, tap any post
4. Tap the site title to go into the site stream (stream no. 1)
5. Tap any post from the site stream
6. Tap the site title to go into the site stream again (stream no. 2)
7. Tap any post from the site stream again
8. Tap the site title to go into the site stream again (stream no. 3)
9. Tap back
10. Tap the site title
11. 🔎 Verify that the app does not crash

### Case 2: Saved post loss in Tags stream

Ensure that you at least follow one tag.

1. Launch the Jetpack app
2. Navigate to Reader
3. Go to the 'Your Tags' stream
4. On one of the tags, scroll to the end and tap 'More'
5. In the Tags stream page, tap the ellipsis button on the first post and tap 'Save'
6. Perform a pull-to-refresh
7. Find the post that you've saved before
8. 🔎 Verify that when tapping the ellipsis menu, the menu item says 'Remove Saved Post' (or the equivalent in your locale)
9. Go to the 'Saved' stream from the stream switcher
10. 🔎 Verify that the post is visible in this stream

## Regression Notes
1. Potential unintended areas of impact
Should be none.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes

4. What automated tests I added (or what prevented me from doing so)
Added unit tests to cover the specific scenarios covered by this PR.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [x] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dvdchr committed:
Fix minor typo

dvdchr committed:
Don't delete previous site topic

There's a cascade delete behavior between ReaderSiteTopic and its
`posts`. When we stack the same site stream several times, the
previous stream's posts will get deleted and raise an exception
because all the managed objects have turned to faults.

We can't also rely on the `inUse` mechanism (i.e., don't delete
if inUse == true), because the stream view controller sets the
topic's `inUse` property to false on deinit--which leads to the
same issue as above.

Note that removing this may reintroduce the bug in #22885, but
I guess not crashing is a better choice for now.

dvdchr committed:
Don't delete saved or in-use posts on tag stream fetch

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/290">25.1</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dvdchr committed:
Add tests for site topic fetch

dvdchr committed:
Add test for ReaderPostStreamService

dvdchr committed:
Fix an issue where the success block could be called twice

When the guard condition is not fulfilled, the success block is
called. A second success block is then called in the `completion`
closure.

dvdchr committed:
Add tests covering non-existent tag

dvdchr commented on WordPress/Classes/Services/ReaderPostStreamService.swift:
```diff
@@ -62,10 +68,15 @@ class ReaderPostStreamService {
-                guard let objectData = object as? NSManagedObject else { continue }
-                context.delete(objectData)
-
-                // Checar se ta em uso ou foi salvo
```
This translates to "_Check if it is in use or has been saved_" 😄

staskus replied:
Hasn't been checked until now apparently 😀

staskus commented on WordPress/Classes/Services/ReaderPostStreamService.swift:
```diff
@@ -18,10 +18,11 @@ class ReaderPostStreamService {
                 guard let readerTopic = try? context.existingObject(with: topic.objectID) as? ReaderAbstractTopic else {
                     // if there was an error or the topic was deleted just bail.
-                    success(0, false)
+                    shouldBail = true
```
Did the previous solution called `success` block twice? Once here and once in the completion block?

dvdchr replied:
Yes, correct. I stumbled upon this while writing tests for this method. I was using `CheckedContinuation` and the continuation got resumed twice.

staskus commented on WordPress/Classes/Services/ReaderPostStreamService.swift:
```diff
@@ -62,10 +68,15 @@ class ReaderPostStreamService {
-                context.delete(objectData)
-
-                // Checar se ta em uso ou foi salvo
+                // do not delete if the post is displayed somewhere or saved by the user.
```
Just to confirm. Is this a completely separate issue from the crash? I was looking at it from the navigation crash perspective as well but that may have been unnecessary.

dvdchr replied:
Yeah, it is a separate issue. I was thinking that both were meant for Reader stability improvements, but they are separate.

staskus commented on WordPress/Classes/Services/ReaderTopicService.m:
```diff
@@ -551,16 +551,7 @@ - (void)siteTopicForSiteWithID:(NSNumber *)siteID
-                [context deleteObject:siteTopic];
-            }
-
+            // always upsert new data to existing site topic.
```
> Note that removing this may reintroduce the bug in https://github.com/wordpress-mobile/WordPress-iOS/pull/22885, but
> I guess not crashing is a better choice for now.

That's a concern.

In the end, I would argue that this crash is a presentation <-> services layer integration issue, coming from a misuse of Core Data. That's not to say that the services layer doesn't have issues or overcomplications but It will be too easy to reintroduce this crash in the presentation layer by making seemingly unrelated improvements within the service.

Are there any other options that you considered? 🤔 For example, could we make sure that we're always using the newest post within a post view rather than holding a Core Data entity with a possibility of invalidation?

dvdchr replied:
After writing that commit, I did a quick test following the steps described in #22885 and it doesn't seem to occur. 

I'll do another pass later and will confirm if this regresses or not. After all, the root cause for #22885 was due to the method always returning the cached object. This commit ensures that the cache is always up-to-date.

staskus replied:
Thanks for letting me know. I'll also retest these regression cases.

staskus reviewed (commented):
Thanks for the work! 

Regarding the crash, could we explore some more options that would not introduce an old issue and would be harder to reintroduce by making changes in the service? Curious what others think as well. IMHO, we could merge the fix during 25.1 beta rather than rush this one.

Regarding the tag stream, as I understand it's a completely separate issue. We could split it into separate PR so it wouldn't be held back by the crash fix.

dvdchr commented:
Thank you for reviewing @staskus ! And good point, although these issues were within the "Reader stability fixes" umbrella, they are solving separate issues.

I'll separate this later!

dvdchr commented:
Closing in favor of #23330 and #23331 .
